magit-bisect-visualize-log: new command

npostavs opened the pull request:
Resolves #2689. Not sure if I chose the best the way to determine args to pass to `magit-log`.

kyleam commented:
> Not sure if I chose the best the way to determine args to pass to magit-log

What alternatives are you considering?  Adding
`magit-bisect-visualize-log` as an action to the log popup and then
aborting if there isn't a bisect in progress?  If so, I prefer the way
you've done it.  And they can be adjusted later by the user from the
`magit-log-refresh-popup`.

kyleam commented on lisp/magit-bisect.el:
```diff
@@ -55,7 +55,8 @@
                         (?r "Reset"        magit-bisect-reset)
-                        (?s "Run script"   magit-bisect-run))
+                        (?s "Run script"   magit-bisect-run)
+                        (?L "Show log"     magit-bisect-visualize-log))
```
Why not lowercase?

npostavs replied:
Because I was already holding down shift for `B` I guess?. Didn't really think much about it.

kyleam commented on lisp/magit-bisect.el:
```diff
@@ -77,6 +78,14 @@ other actions from the bisect popup (\
+  "Show log of remaining commits to bisect."
+  (interactive)
+  (magit-log '("--bisect") (default-value 'magit-log-arguments)
+             (split-string (magit-file-line (magit-git-dir "BISECT_NAMES"))
```
4bbc8d6eb9e0afa54d15422167d49e4eea2a7f41

npostavs commented:
> What alternatives are you considering?

I didn't have anything else in mind, this is just the first thing that seemed to work after I realized that just calling `(magit-log-argument)` wasn't right. If it looks reasonable to you, I guess that's a good sign. :)

kyleam commented:
I wonder if it'd be better to use a locked buffer so that it doesn't
interfere with the main log buffer. `magit-bisect-visualize-log`
could be changed to something like

    (magit-mode-setup-internal
           #'magit-log-mode
           <same args>
           t)

kyleam commented:
> split-string isn't enough either to quote cover names with internal single quotes though

Shoot, I should've stuck to just pointing out the issue :)

npostavs commented:
> I wonder if it'd be better to use a locked buffer so that it doesn't
> interfere with the main log buffer.

Yeah I think it makes sense. I also changed the arg grabbing to be like `magit-log-buffer-file`.

>> split-string isn't enough either to quote cover names with internal single quotes though
>
>Shoot, I should've stuck to just pointing out the issue :)

Added a function to handle it correctly. It feels a bit like overkill though...

npostavs commented:
> Added a function to handle it correctly. It feels a bit like overkill though...

The overkill was annoying me, I changed to shelling out to git instead.

kyleam commented:
nice

npostavs commented:
Ok, I've rebased this on master, only question mark left for me is the keybinding of `magit-bisect-visualize-log`, `L` or `l`?

kyleam commented:
As my comment suggests, I prefer `l`, but I don't feel strongly about it.

pvinis commented:
besides `l` and `L`, is there something else keeping that from merging? i think i would also prefer `l`, since `g`, `b` etc are lower case.

npostavs commented:
> besides `l` and `L`, is there something else keeping that from merging?

Now that I about it, documentation in the manual.

> i think i would also prefer `l`, since `g`, `b` etc are lower case.

Ok, I'll change to `l` then.

tarsius commented:
Nice job!

However I am not sure we really need all of this.

A simpler approach would be to simply teach `magit-insert-bisect-rest` to use "--graph" if it is a member of `magit-log-section-arguments`. Or do you expect that there would be situations where the use of a graph would make it harder to reason about the ongoing bisect? If so then it might be better to add a separate `magit-bisect-show-graph` option and a toggle command instead of depending on `magit-log-section-arguments`.

pvinis commented:
that sounds nice. a toggle in the bisect popup, or just getting the graph option from log, would be great. i think @npostavs just put it in a different function so we can try it out.

npostavs commented:
Ah, I had forgotten about `magit-log-section-arguments`. Yeah, probably makes sense to use that.

> Or do you expect that there would be situations where the use of a graph would make it harder to reason about the ongoing bisect?

Probably the opposite, i.e. the graph should make it easier to see what's going on. But maybe one might want to avoid adding the `--graph` to `magit-log-section-arguments` for performance reasons, while still wanting it during bisection?

tarsius commented:
I went with adding `magit-bisect-show-graph` and defaulting to `t`.

tarsius commented on lisp/magit-utils.el:
```diff
@@ -412,6 +412,11 @@ If optional NUM is specified only delete that subexpression."
   (delete-region (match-beginning (or num 0))
                  (match-end (or num 0))))
 
+(defun magit-split-and-unsq-quote-string (str)
```
Might this still be useful elsewhere?

npostavs replied:
In theory yes, though I don't know of any place that would currently use it.
